eth_utils/conversions.py

```diff
@@ -115,13 +122,14 @@ def to_text(primitive=None, hexstr=None, text=None):
     raise TypeError("Expected an int, bytes, bytearray or hexstr.")
 
 
-def text_if_str(to_type, text_or_primitive):
+def text_if_str(to_type: Callable[..., Union[bytes, int, str]],
+                text_or_primitive: Union[bytes, int, str]) -> Union[bytes, int, str]:
```

Comment:
I think the only way for us to type these functions is using generics.

https://docs.python.org/3/library/typing.html#generics

Basically, `to_type` will be something like `Callable[[Optional[Primative], Optional[str], Optional[str]], Generic[T]` where `T = TypeVar('T')`.  Then `hexstr_if_hex` should be marked as returning type `T`.  This may not be exactly right, but the idea is that we setup a generic type `T` which `to_type` returns and say that `hexstr_if_hex` will always return the same type as `to_type`.